docs/git/cheat-sheet.md

```diff
@@ -0,0 +1,67 @@
+# Git Cheat Sheet
+
+The following is a cheat sheet for using [git] while developing Cardano SL
+which should contain all the git operations that someone working on Cardano
+will need at some point.
+
+### Cloning the Cardano SL repository
+```
+git clone https://github.com/input-output-hk/cardano-sl
+```
+After you change directory into that cloned repository, you may want to tell
+git to use the SSH protocol for the "git push" command, keeping the HTTPS
+protocol for the "git pull" and "git fetch" commands by issuing the following
+command:
+
+```
+git remote set-url --push origin git@github.com:input-output-hk/cardano-sl
+```
+
+### Listing the current and local branches
+```
+git branch
+```
+To list all the branches both local and at the origin (usually Github):
+```
+git branch --all
+```
+
+### Create and change into a new branch
+Development on Cardano SL is usually done by branching off the "develop" branch.
+Assuming you are already on the develop branch and your user name is "xyz" and
+you are working on YouTrack ticket number "YT-1234":
+```
+git branch xyz/YT-1234
+git checkout xyz/YT-1234
+```
+A shorter version of the above is:
+```
+git checkout -b xyz/YT-1234
+```
+
+### Rebasing your changes against develop
+While you are working on your branch, it is not uncommon for other developers
+to merge their changes into the develop branch. To pull those changes into
+your branch:
+```
+git fetch origin
+git rebase origin/develop
+```
```

Comment:
I usually always add two options two a rebase command:

- `--preserve-merges`: without which git will _"flatten"_ the history and remove all merge commits. This is really bad for keeping a nice a clear history, especially if we have a branch-based work-flow.

```
* c835fde                                                                     
* 1d9dbe9                                                                        
|\                              * c835fde                                                  
| * 66f8430                     * 1d9dbe9                                                         
| * 7e57872                     * 66f8430                                                         
|/                              * 7e57872                                                  
* b18da81                       * b18da81                                                      
|\                              * 963c914                                                  
| * 963c914         ----->      * e88880b                                                        
| * e88880b                     * a0008aa                                                        
| |                             * 2261860                                                  
* | a0008aa                     * b20ea60                                                         
|/                                                                         
|                                                                          
* 2261860                                                                    
* b20ea60                                                                    


BEFORE REBASE       ----->     AFTER REBASE
```

- `--committer-date-is-author-date`: Rather optional but it prevents the commit's date to be changed by the "rebase's date". I like to preserve time this way but this is not always possible.

```
git rebase --preserve-merges --committer-date-is-author-date origin/develop
```

Note that default options can be tweaked via the git configuration. e.g.:

```
[rebase]
  preserveMerges = true
```